# Task 4.5 Data Consistency Checks

# Table of contents
#### 1. Importing Libraries
#### 2. Importing Data
#### 3. Data Consistency Checks
#### 4. Task 4.5

# 01. Importing Libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Turn project folder path into a string

path = r'C:\Users\osoty\Desktop\CareerFoundry Data Analytics\Data Immersion\Achievement 4\02-23 Instacart Basket Analysis'

In [3]:
# Import the "products.csv" data set from the "Original Data" folder

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import the "orders_wrangled.csv" data set from the "Prepared Data" folder

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data Consistency Checks

## Mixed Data Types

In [5]:
# Create a test dataframe

df_test = pd.DataFrame()

In [6]:
# Create a mixed column type

df_test['mix'] = ['a', 'b', 1, True]

In [7]:
# Check for mixed data types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [8]:
# Change mix to string type

df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [9]:
# Find missing values in all columns

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# List the missing values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
# Create a new dataframe without the missing rows

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [12]:
# Check the current number of rows in the dataframe

df_prods_clean.shape

(49677, 5)

## Duplicates

In [13]:
# Look for full duplicates within the dataframe

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
# Check the current number of rows within the dataframe

df_prods_clean.shape

(49677, 5)

In [16]:
# Create a new dataframe without the duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
# Check the current number of rows within the new dataframe

df_prods_clean_no_dups.shape

(49672, 5)

# 04. Task 4.5

In [18]:
# 1. Data Consistency Checks df_prods dataframe - Descriptive Statistics
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

### The maximum value for prices is 99999 which seems off. This could either be an error or the number 99999 could be a placeholder for missing values.

In [19]:
# Find out what items have a price listed as 99999

df_prods_clean_no_dups[df_prods_clean_no_dups.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### The price for 2% Reduced Fat Milk is obviously an error or placeholder.

In [20]:
# 2. Check the df_ords dataframe

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [21]:
df_ords.shape

(3421083, 7)

### There  is an additional column "Unnamed: 0" which is repeating the values for order_id and was probably created while exporting the file or importing it. Besides that everything looks good.

In [22]:
# Drop the "Unnamed" column for df_ords

df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [23]:
# Check if the additional column was dropped

df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [24]:
# 3. Check for mixed-typed data in your df_ords dataframe

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)
    

### No mixed-typed  data was found.

In [25]:
# 5. Check for missing values in your df_ords dataframe

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### There are 206209 missing values for the "days_since_prior_order" column.

In [26]:
df_ords.head(15)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   1187899        1            11                  4                  8   
11   2168274        2             1                  2                 11   
12   1501582        2             2                  5                 10   
13   1901567        2             3                  1                 10   
14    738281        2             4                  2                 10   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0

In [27]:
# List the missing values

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [28]:
df_ords_nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

   ### Looking at df_ords_nan shows that all observations with a NaN for days_since_prior_order also have an order_number of 1 which means it is the first order for each customer. The max user_id is 206209 which corresponds to the number of missing values, meaning if a customer is ordering for the first time, there won't be a value for days_since_prior_order.

In [29]:
# 6. Address the missing values:

# Create a new dataframe and set a new column for user

df_ords_clean = df_ords

In [30]:
# Create a new column for flagging a new customer

df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull()==True

In [31]:
# Check if the new column is there

df_ords_clean

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]

In [32]:
df_ords_clean.shape

(3421083, 7)

### I decided to create a flag in form of a new column named "new_customer".

In [33]:
# 7. Check for duplicates in df_ords 

df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [34]:
# View df_ords_dups

df_ords_dups.size

0

In [35]:
df_ords_clean.shape

(3421083, 7)

### No duplicates were found.

In [36]:
# 9. Export final cleaned data:

# Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

# Export df_ords_dups

df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)
